controller/src/main/java/io/pravega/controller/store/stream/Scope.java

```diff
@@ -39,6 +44,35 @@
      */
     CompletableFuture<Void> deleteScope();
 
+    /**
```

Comment:
I don't understand these two API calls being added here. Why do we need to add `addStreamToScope` and `removeStreamFromScope` here for listing streams?

Reply:
The reason we are doing this is so that we allow store to keep list of stream grouping separate from stream metadata for all streams. 
This allows us to break down list of streams such that we can group them and bring in a continuation token based logic. 

For example, in zookeeper based store, rest of the stream metadata is structured as /<scope>/<stream>
If we were to list all streams in scope, we only had the option of fetching all children of `/<scope>`. 
And if we had to restructure stream metadata into different sets, it would have been backward incompatible. 
So i simply introduced these new apis which allows `store` to group streams under scope which is optimal for listStreamsInScope. 
So the stream metadata continues to be under `/<scope>/<stream>` but stream identifiers are also stored under a new znode hierarchy `/streamsInScope/scope/<set>/<stream>->creationTime`. So all streams under the scope are now divided under these predetermined number of sets. 
So a listStreamsInScope starts from a continuation token (initially null/empty) from which it extracts set and last record identifier. 
So when we do getChildren on `/streamsInScope/scope/<set>` we fetch an order of magnitude smaller number of streams in one query. And only take the limited number of entries that are ahead of continuation token. 
If we found less than `limit` number of entries from one set, we move to the next set. so on and so forth.

Reply:
This observation made me thing of the hierarchical ledger manager scheme we introduced in BK to solve precisely this problem, see Hierarchical Ledger Manager here:

https://bookkeeper.apache.org/docs/4.5.0/getting-started/concepts/

Can we apply this concept of a hierarchy based on ID here?

Reply:
this looks good too. I will make the change to incorporate this. it will take care of several nuances for us. 
the only thing is we will need to store this "path" under stream's metadata too which is not necessarily a problem.. but adds another step in stream creation and its handling for failures.. 

And well its a neat idea that takes care of some nuances with the current scheme.